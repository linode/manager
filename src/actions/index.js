export { setTitle } from './title';
export { setSource } from './source';
export { setAnalytics } from './analytics';
