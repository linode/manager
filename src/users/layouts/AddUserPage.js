import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router';
import { push } from 'react-router-redux';

import { Card, CardHeader } from 'linode-components/cards';
import { reduceErrors } from '~/errors';
import { users } from '~/api';
import { UserForm } from '../components/UserForm';

export class AddUserPage extends Component {
  constructor() {
    super();

    this.onSubmit = this.onSubmit.bind(this);
    this.state = {
      loading: false,
      errors: {},
    };
  }

  async onSubmit(stateValues) {
    const { dispatch } = this.props;

    this.setState({ loading: true });
    try {
      await dispatch(users.post(stateValues));
      dispatch(push('/users'));
    } catch (response) {
      const errors = await reduceErrors(response);
      this.setState({ errors, loading: false });
    }
  }

  render() {
    const { errors } = this.state;
    return (
      <div className="PrimaryPage container">
        <Link to="/users">Users</Link>
        <header className="PrimaryPage-header">
          <div className="PrimaryPage-headerRow clearfix">
            <h1>Users</h1>
          </div>
        </header>
        <div className="PrimaryPage-body User-create">
          <Card header={<CardHeader title="Add a user" />}>
            <UserForm
              errors={errors}
              onSubmit={this.onSubmit}
            />
          </Card>
        </div>
      </div>
    );
  }
}

AddUserPage.propTypes = {
  dispatch: PropTypes.func.isRequired,
};

function select() {
  return {};
}

export default connect(select)(AddUserPage);
