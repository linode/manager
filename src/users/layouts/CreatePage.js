import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router';

import { Card } from 'linode-components/cards';

import { UserForm } from '../components';


export function CreatePage(props) {
  const { dispatch } = props;

  return (
    <div className="PrimaryPage container">
      <Link to="/users">Users</Link>
      <header className="PrimaryPage-header">
        <div className="PrimaryPage-headerRow clearfix">
          <h1>Add a user</h1>
        </div>
      </header>
      <div className="PrimaryPage-body">
        <Card>
          <UserForm dispatch={dispatch} />
        </Card>
      </div>
    </div>
  );
}

CreatePage.propTypes = {
  dispatch: PropTypes.func.isRequired,
};

export default connect()(CreatePage);
