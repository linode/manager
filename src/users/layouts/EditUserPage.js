import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { push } from 'react-router-redux';

import { setError } from '~/actions/errors';
import { users } from '~/api';
import { Card } from '~/components/cards';
import { reduceErrors } from '~/errors';
import { UserForm } from '../components/UserForm';

export class EditUserPage extends Component {
  static async preload({ dispatch }, { username }) {
    try {
      await dispatch(users.one([username]));
    } catch (response) {
      // eslint-disable-next-line no-console
      console.error(response);
      dispatch(setError(response));
    }
  }

  constructor(props) {
    super(props);

    this.onSubmit = this.onSubmit.bind(this);
    const { username } = props.params;
    this.state = {
      loading: false,
      initUsername: username,
      email: props.users[username].email,
      errors: {},
    };
  }

  async onSubmit(stateValues) {
    const { dispatch } = this.props;
    const { initUsername } = this.state;

    this.setState({ loading: true });
    try {
      await dispatch(users.put(stateValues, initUsername));
      dispatch(push('/users'));
    } catch (response) {
      const errors = await reduceErrors(response);
      this.setState({ errors, loading: false });
    }
  }

  render() {
    const { initUsername, email, errors } = this.state;
    return (
      <div className="PrimaryPage container">
        <header className="PrimaryPage-header">
          <div className="PrimaryPage-headerRow clearfix">
            <h1>Users</h1>
          </div>
        </header>
        <div className="PrimaryPage-body User-edit">
          <Card title={`Edit ${initUsername}`}>
            <UserForm
              username={initUsername}
              email={email}
              errors={errors}
              onSubmit={this.onSubmit}
            />
          </Card>
        </div>
      </div>
    );
  }
}

EditUserPage.propTypes = {
  dispatch: PropTypes.func.isRequired,
  params: PropTypes.shape({
    username: PropTypes.string,
  }),
  users: PropTypes.object.isRequired,
};

function select(state) {
  return {
    users: state.api.users.users,
  };
}

export default connect(select)(EditUserPage);
