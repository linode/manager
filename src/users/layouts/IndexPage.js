import _ from 'lodash';
import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router';

import { setSource } from '~/actions/source';
import { setTitle } from '~/actions/title';
import { users } from '~/api';

import { User } from '../components';


export class IndexPage extends Component {
  static async preload({ dispatch }) {
    await dispatch(users.all());
  }

  async componentDidMount() {
    const { dispatch } = this.props;
    dispatch(setSource(__filename));

    dispatch(setTitle('Users'));
  }

  renderGroup = (group, i, groups) => {
    const { username: currentUser } = this.props.profile;
    const _renderGroup = group.map(user => (
      <div className="col-lg-6" key={user.username}>
        <User user={user} currentUser={currentUser} />
      </div>
    ));

    if (i === groups.length - 1) {
      return <div className="row" key={i}>{_renderGroup}</div>;
    }

    return <section className="row" key={i}>{_renderGroup}</section>;
  }

  render() {
    const listOfUsers = Object.values(this.props.users.users);

    return (
      <div className="PrimaryPage container">
        <header className="PrimaryPage-header">
          <div className="PrimaryPage-headerRow clearfix">
            <h1 className="float-sm-left">Users</h1>
            <Link to="/users/create" className="linode-add btn btn-primary float-sm-right">
              <span className="fa fa-plus"></span>
              Add a user
            </Link>
          </div>
        </header>
        <div className="PrimaryPage-body">
          {_.chunk(listOfUsers, 2).map(this.renderGroup)}
        </div>
      </div>
    );
  }
}

IndexPage.propTypes = {
  dispatch: PropTypes.func,
  users: PropTypes.object,
  profile: PropTypes.object,
};

function select(state) {
  return {
    users: state.api.users,
    profile: state.api.profile,
  };
}

export default connect(select)(IndexPage);
