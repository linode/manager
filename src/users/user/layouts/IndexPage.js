import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router';
import { push } from 'react-router-redux';

import { Tabs } from 'linode-components/tabs';

import { getObjectByLabelLazily } from '~/api/util';

import { setTitle } from '~/actions/title';


export class IndexPage extends Component {
  static async preload({ dispatch, getState }, { username }) {
    await dispatch(getObjectByLabelLazily('users', username, 'username'));
  }

  async componentDidMount() {
    const { dispatch, user } = this.props;
    dispatch(setTitle(user.username));
  }

  render() {
    if (!this.props.user) {
      return null;
    }

    const { username, restricted } = this.props.user;
    const tabList = [{ name: 'Dashboard', link: '' }];
    if (restricted) {
      tabList.push({ name: 'Permissions', link: '/permissions' });
    }
    const tabs = tabList.map(t => ({ ...t, link: `/users/${username}${t.link}` }));

    return (
      <div>
        <header className="main-header">
          <div className="container">
            <div className="float-sm-left">
              <Link to="/users">Users</Link>
              <h1 title={username}>
                <Link to={`/users/${username}`}>
                  {username}
                </Link>
              </h1>
            </div>
          </div>
        </header>
        <div className="main-header-fix"></div>
        <Tabs
          tabs={tabs}
          onClick={(e, tabIndex) => {
            e.stopPropagation();
            this.props.dispatch(push(tabs[tabIndex].link));
          }}
          pathname={location.pathname}
        >
          {this.props.children}
        </Tabs>
      </div>
    );
  }
}

IndexPage.propTypes = {
  dispatch: PropTypes.func,
  user: PropTypes.object,
  children: PropTypes.node,
};

export function selectUser(state, props) {
  const { users } = state.api.users;
  const { username } = props.params;
  const user = users[username];
  return { user };
}

export default connect(selectUser)(IndexPage);
