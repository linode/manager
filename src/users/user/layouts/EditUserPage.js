import React, { PropTypes } from 'react';
import { connect } from 'react-redux';

import { Card } from 'linode-components/cards';

import { selectUser } from './IndexPage';
import { UserForm } from '../../components';


export function EditUserPage(props) {
  const { user, dispatch } = props;

  return (
    <Card>
      <UserForm user={user} dispatch={dispatch} />
    </Card>
  );
}

EditUserPage.propTypes = {
  user: PropTypes.object,
  dispatch: PropTypes.func.isRequired,
};

export default connect(selectUser)(EditUserPage);
