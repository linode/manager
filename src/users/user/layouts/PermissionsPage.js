import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';

import { Card } from 'linode-components/cards';
import {
  Checkbox,
  Checkboxes,
  Form,
  FormGroup,
  FormSummary,
  SubmitButton,
} from 'linode-components/forms';

import { users } from '~/api';
import { dispatchOrStoreErrors, getObjectByLabelLazily } from '~/api/util';

import { selectUser } from './IndexPage';
import { PermissionsTable } from '../components';


export class PermissionsPage extends Component {
  static async preload({ dispatch, getState }, { username }) {
    const user = await dispatch(getObjectByLabelLazily('users', username, 'username'));

    if (user.restricted) {
      await dispatch(users.permissions.one([username]));
    }
  }

  constructor(props) {
    super(props);

    this.state = {
      ...props.user._permissions,
      loading: false,
      errors: {},
    };
  }

  onSubmit = () => {
    const { dispatch, user: { username } } = this.props;
    const { global, customer, linode, nodebalancer, domain } = this.state;
    const data = { global, customer, linode, nodebalancer, domain };

    return dispatch(dispatchOrStoreErrors.call(this, [
      () => users.permissions.put(data, username),
    ]));
  }

  onCellChange = (record, checked, keys) => {
    const { parentKey, dataKey } = keys;
    const parentState = this.state[parentKey].map(function (child) {
      if (child === record) {
        return {
          ...child,
          [dataKey]: checked,
        };
      }
      return child;
    });

    this.setState({
      [parentKey]: parentState,
    });
  }

  updateGlobal = (name) => {
    this.setState({
      global: {
        ...this.state.global,
        [name]: !this.state.global[name],
      },
    });
  }

  render() {
    const { global, customer, linode, domain, nodebalancer, loading, errors } = this.state;

    return (
      <Form onSubmit={this.onSubmit}>
        <Card className="Permissions">
          <div className="Permissions-section">
            <h3>Global Permissions</h3>
            <Checkboxes>
              <Checkbox
                id="permission-customer-access"
                checked={customer.access}
                onChange={() => this.setState({
                  customer: {
                    ...this.state.customer,
                    access: !customer.access,
                  },
                })}
                label="Can view invoices, make payments, update contact and
                  billing info, and will receive copies of all invoices and
                  payment emails"
              />
              <Checkbox
                id="permission-customer-cancel"
                checked={customer.cancel}
                onChange={() => this.setState({
                  customer: {
                    ...this.state.customer,
                    cancel: !customer.cancel,
                  },
                })}
                label="Can cancel the entire account"
              />
              <Checkbox
                id="permission-global-linodes"
                checked={global.add_linodes}
                onChange={() => this.updateGlobal('add_linodes')}
                label="Can add Linodes to this Account ($)"
              />
              <Checkbox
                id="permission-global-nodebalancers"
                checked={global.add_nodebalancers}
                onChange={() => this.updateGlobal('add_nodebalancers')}
                label="Can add NodeBalancers to this Account ($)"
              />
              <Checkbox
                id="permission-global-domains"
                checked={global.add_domains}
                onChange={() => this.updateGlobal('add_domains')}
                label="Can add Domains"
              />
            </Checkboxes>
          </div>
          <PermissionsTable
            title="Linode"
            parentKey="linode"
            onCellChange={this.onCellChange}
            objects={linode}
            columns={[
              { dataKey: 'all', label: 'All' },
              { dataKey: 'access', label: 'Access' },
              { dataKey: 'delete', label: 'Delete' },
              { dataKey: 'resize', label: 'Resize ($)' },
            ]}
          />
          <PermissionsTable
            title="NodeBalancer"
            parentKey="nodebalancer"
            onCellChange={this.onCellChange}
            objects={nodebalancer}
            columns={[
              { dataKey: 'all', label: 'All' },
              { dataKey: 'access', label: 'Access' },
              { dataKey: 'delete', label: 'Delete' },
              null,
            ]}
          />
          <PermissionsTable
            title="Domains"
            parentKey="domain"
            onCellChange={this.onCellChange}
            objects={domain}
            columns={[
              { dataKey: 'all', label: 'All' },
              { dataKey: 'access', label: 'Access' },
              { dataKey: 'delete', label: 'Delete' },
              null,
            ]}
          />
          <FormGroup>
            <SubmitButton disabled={loading}>Save</SubmitButton>
            <FormSummary errors={errors} success="Permissions saved." />
          </FormGroup>
        </Card>
      </Form>
    );
  }
}

PermissionsPage.propTypes = {
  user: PropTypes.object.isRequired,
  dispatch: PropTypes.func.isRequired,
};

export default connect(selectUser)(PermissionsPage);
