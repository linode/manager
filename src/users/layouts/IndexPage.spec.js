import { mount, shallow } from 'enzyme';
import React from 'react';
import sinon from 'sinon';
import { StaticRouter } from 'react-router-dom';

import { SHOW_MODAL } from '~/actions/modal';
import { getEmailHash } from '~/cache';
import { GRAVATAR_BASE_URL } from '~/constants';
import { IndexPage } from '~/users/layouts/IndexPage';

import { api } from '~/data';
import { expectRequest } from '~/test.helpers.js';
import { profile } from '~/data/profile';


const { users } = api;

describe('users/layouts/IndexPage', () => {
  const sandbox = sinon.sandbox.create();

  afterEach(() => {
    sandbox.restore();
  });

  const dispatch = sandbox.spy();

  it('should render without error', () => {
    const mockDispatch = jest.fn();
    const wrapper = shallow(
      <StaticRouter>
        <IndexPage dispatch={mockDispatch} selectedMap={{}} profile={profile} users={users} />
      </StaticRouter>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a list of Users', () => {
    const page = mount(
      <StaticRouter>
        <IndexPage
          dispatch={dispatch}
          selectedMap={{}}
          profile={profile}
          users={users}
        />
      </StaticRouter>
    );

    const zone = page.find('.TableRow');
    // + 1 for the group
    expect(zone.length).toBe(Object.keys(users.users).length);
    const firstZone = zone.at(0);
    expect(firstZone.find('td img').props().src)
      .toBe(`${GRAVATAR_BASE_URL}${getEmailHash('example1@domain.com')}`);
    expect(firstZone.find('Link').props().to)
      .toBe('/users/testuser1');
    expect(firstZone.find('td').at(2).text())
      .toBe('testuser1');
    expect(firstZone.find('td').at(3).text())
      .toBe('example1@domain.com');
    expect(firstZone.find('td').at(4).text())
      .toBe('Unrestricted');
  });

  it('shows the delete modal when delete is pressed', () => {
    const page = mount(
      <StaticRouter>
        <IndexPage
          dispatch={dispatch}
          selectedMap={{}}
          profile={profile}
          users={users}
        />
      </StaticRouter>
    );

    const zoneDelete = page.find('.TableRow Button').at(0);
    dispatch.reset();
    zoneDelete.simulate('click');
    expect(dispatch.callCount).toBe(1);
    expect(dispatch.firstCall.args[0]).toHaveProperty('type');
    expect(dispatch.firstCall.args[0].type).toBe(SHOW_MODAL);
  });

  it.skip('deletes selected users when delete is pressed', async () => {
    const page = mount(
      <StaticRouter>
        <IndexPage
          dispatch={dispatch}
          selectedMap={{ 1: true }}
          profile={profile}
          users={users}
        />
      </StaticRouter>
    );

    dispatch.reset();

    page.find('tr button').at(0).simulate('click');
    expect(dispatch.callCount).toBe(1);
    const modal = mount(dispatch.firstCall.args[0].body);

    dispatch.reset();
    modal.find('Form').props().onSubmit({ preventDefault() { } });
    const fn = dispatch.firstCall.args[0];
    await expectRequest(fn, '/account/users/testuser1', { method: 'DELETE' });
  });
});
