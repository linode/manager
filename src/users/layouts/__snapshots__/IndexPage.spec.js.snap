// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users/layouts/IndexPage should render without error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <StaticRouter
    basename=""
    location="/"
  >
    <IndexPage
      dispatch={[Function]}
      profile={
        Object {
          "authorized_keys": Array [
            "foobar",
          ],
          "email": "user@example.com",
          "email_notifications": true,
          "ip_whitelist_enabled": false,
          "lish_auth_method": "password_keys",
          "referrals": Object {
            "code": "",
            "completed": 0,
            "credit": 0,
            "pending": 0,
            "total": 0,
            "url": null,
          },
          "restricted": false,
          "timezone": "US/Eastern",
          "two_factor_auth": true,
          "username": "theuser",
        }
      }
      selectedMap={Object {}}
      users={
        Object {
          "ids": Array [
            undefined,
            undefined,
          ],
          "pagesFetched": Array [
            1,
          ],
          "plural": "users",
          "singular": "user",
          "totalPages": 1,
          "totalResults": 2,
          "users": Array [
            Object {
              "_polling": false,
              "email": "example1@domain.com",
              "password": "crisprcas9",
              "restricted": false,
              "username": "testuser1",
            },
            Object {
              "_permissions": Object {
                "domain": Array [
                  Object {
                    "id": 3333,
                    "label": "domain1",
                    "permissions": "read_write",
                  },
                ],
                "global": Object {
                  "account_access": "read_only",
                  "add_domains": true,
                  "add_images": true,
                  "add_linodes": true,
                  "add_longview": true,
                  "add_nodebalancers": true,
                  "add_stackscripts": true,
                  "add_volumes": true,
                  "cancel_account": false,
                  "longview_subscription": true,
                },
                "image": Array [
                  Object {
                    "id": 6666,
                    "label": "image1",
                    "permissions": "read_only",
                  },
                  Object {
                    "id": 7777,
                    "label": "image2",
                    "permissions": "read_write",
                  },
                ],
                "linode": Array [
                  Object {
                    "id": 1111,
                    "label": "linode1",
                    "permissions": "read_write",
                  },
                ],
                "longview": Array [
                  Object {
                    "id": 5555,
                    "label": "longview1",
                    "permissions": "read_only",
                  },
                ],
                "nodebalancer": Array [
                  Object {
                    "id": 2222,
                    "label": "nb1",
                    "permissions": "read_only",
                  },
                ],
                "stackscript": Array [
                  Object {
                    "id": 4444,
                    "label": "stackscript1",
                    "permissions": null,
                  },
                ],
                "volume": Array [
                  Object {
                    "id": 8888,
                    "label": "volume1",
                    "permissions": null,
                  },
                ],
              },
              "_polling": false,
              "email": "example2@domain.com",
              "password": "crisprcas9",
              "restricted": true,
              "username": "testuser2",
            },
          ],
        }
      }
    />
  </StaticRouter>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <IndexPage
        dispatch={[Function]}
        profile={
          Object {
            "authorized_keys": Array [
              "foobar",
            ],
            "email": "user@example.com",
            "email_notifications": true,
            "ip_whitelist_enabled": false,
            "lish_auth_method": "password_keys",
            "referrals": Object {
              "code": "",
              "completed": 0,
              "credit": 0,
              "pending": 0,
              "total": 0,
              "url": null,
            },
            "restricted": false,
            "timezone": "US/Eastern",
            "two_factor_auth": true,
            "username": "theuser",
          }
        }
        selectedMap={Object {}}
        users={
          Object {
            "ids": Array [
              undefined,
              undefined,
            ],
            "pagesFetched": Array [
              1,
            ],
            "plural": "users",
            "singular": "user",
            "totalPages": 1,
            "totalResults": 2,
            "users": Array [
              Object {
                "_polling": false,
                "email": "example1@domain.com",
                "password": "crisprcas9",
                "restricted": false,
                "username": "testuser1",
              },
              Object {
                "_permissions": Object {
                  "domain": Array [
                    Object {
                      "id": 3333,
                      "label": "domain1",
                      "permissions": "read_write",
                    },
                  ],
                  "global": Object {
                    "account_access": "read_only",
                    "add_domains": true,
                    "add_images": true,
                    "add_linodes": true,
                    "add_longview": true,
                    "add_nodebalancers": true,
                    "add_stackscripts": true,
                    "add_volumes": true,
                    "cancel_account": false,
                    "longview_subscription": true,
                  },
                  "image": Array [
                    Object {
                      "id": 6666,
                      "label": "image1",
                      "permissions": "read_only",
                    },
                    Object {
                      "id": 7777,
                      "label": "image2",
                      "permissions": "read_write",
                    },
                  ],
                  "linode": Array [
                    Object {
                      "id": 1111,
                      "label": "linode1",
                      "permissions": "read_write",
                    },
                  ],
                  "longview": Array [
                    Object {
                      "id": 5555,
                      "label": "longview1",
                      "permissions": "read_only",
                    },
                  ],
                  "nodebalancer": Array [
                    Object {
                      "id": 2222,
                      "label": "nb1",
                      "permissions": "read_only",
                    },
                  ],
                  "stackscript": Array [
                    Object {
                      "id": 4444,
                      "label": "stackscript1",
                      "permissions": null,
                    },
                  ],
                  "volume": Array [
                    Object {
                      "id": 8888,
                      "label": "volume1",
                      "permissions": null,
                    },
                  ],
                },
                "_polling": false,
                "email": "example2@domain.com",
                "password": "crisprcas9",
                "restricted": true,
                "username": "testuser2",
              },
            ],
          }
        }
      />,
      "history": Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "dispatch": [Function],
        "profile": Object {
          "authorized_keys": Array [
            "foobar",
          ],
          "email": "user@example.com",
          "email_notifications": true,
          "ip_whitelist_enabled": false,
          "lish_auth_method": "password_keys",
          "referrals": Object {
            "code": "",
            "completed": 0,
            "credit": 0,
            "pending": 0,
            "total": 0,
            "url": null,
          },
          "restricted": false,
          "timezone": "US/Eastern",
          "two_factor_auth": true,
          "username": "theuser",
        },
        "selectedMap": Object {},
        "users": Object {
          "ids": Array [
            undefined,
            undefined,
          ],
          "pagesFetched": Array [
            1,
          ],
          "plural": "users",
          "singular": "user",
          "totalPages": 1,
          "totalResults": 2,
          "users": Array [
            Object {
              "_polling": false,
              "email": "example1@domain.com",
              "password": "crisprcas9",
              "restricted": false,
              "username": "testuser1",
            },
            Object {
              "_permissions": Object {
                "domain": Array [
                  Object {
                    "id": 3333,
                    "label": "domain1",
                    "permissions": "read_write",
                  },
                ],
                "global": Object {
                  "account_access": "read_only",
                  "add_domains": true,
                  "add_images": true,
                  "add_linodes": true,
                  "add_longview": true,
                  "add_nodebalancers": true,
                  "add_stackscripts": true,
                  "add_volumes": true,
                  "cancel_account": false,
                  "longview_subscription": true,
                },
                "image": Array [
                  Object {
                    "id": 6666,
                    "label": "image1",
                    "permissions": "read_only",
                  },
                  Object {
                    "id": 7777,
                    "label": "image2",
                    "permissions": "read_write",
                  },
                ],
                "linode": Array [
                  Object {
                    "id": 1111,
                    "label": "linode1",
                    "permissions": "read_write",
                  },
                ],
                "longview": Array [
                  Object {
                    "id": 5555,
                    "label": "longview1",
                    "permissions": "read_only",
                  },
                ],
                "nodebalancer": Array [
                  Object {
                    "id": 2222,
                    "label": "nb1",
                    "permissions": "read_only",
                  },
                ],
                "stackscript": Array [
                  Object {
                    "id": 4444,
                    "label": "stackscript1",
                    "permissions": null,
                  },
                ],
                "volume": Array [
                  Object {
                    "id": 8888,
                    "label": "volume1",
                    "permissions": null,
                  },
                ],
              },
              "_polling": false,
              "email": "example2@domain.com",
              "password": "crisprcas9",
              "restricted": true,
              "username": "testuser2",
            },
          ],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <IndexPage
          dispatch={[Function]}
          profile={
            Object {
              "authorized_keys": Array [
                "foobar",
              ],
              "email": "user@example.com",
              "email_notifications": true,
              "ip_whitelist_enabled": false,
              "lish_auth_method": "password_keys",
              "referrals": Object {
                "code": "",
                "completed": 0,
                "credit": 0,
                "pending": 0,
                "total": 0,
                "url": null,
              },
              "restricted": false,
              "timezone": "US/Eastern",
              "two_factor_auth": true,
              "username": "theuser",
            }
          }
          selectedMap={Object {}}
          users={
            Object {
              "ids": Array [
                undefined,
                undefined,
              ],
              "pagesFetched": Array [
                1,
              ],
              "plural": "users",
              "singular": "user",
              "totalPages": 1,
              "totalResults": 2,
              "users": Array [
                Object {
                  "_polling": false,
                  "email": "example1@domain.com",
                  "password": "crisprcas9",
                  "restricted": false,
                  "username": "testuser1",
                },
                Object {
                  "_permissions": Object {
                    "domain": Array [
                      Object {
                        "id": 3333,
                        "label": "domain1",
                        "permissions": "read_write",
                      },
                    ],
                    "global": Object {
                      "account_access": "read_only",
                      "add_domains": true,
                      "add_images": true,
                      "add_linodes": true,
                      "add_longview": true,
                      "add_nodebalancers": true,
                      "add_stackscripts": true,
                      "add_volumes": true,
                      "cancel_account": false,
                      "longview_subscription": true,
                    },
                    "image": Array [
                      Object {
                        "id": 6666,
                        "label": "image1",
                        "permissions": "read_only",
                      },
                      Object {
                        "id": 7777,
                        "label": "image2",
                        "permissions": "read_write",
                      },
                    ],
                    "linode": Array [
                      Object {
                        "id": 1111,
                        "label": "linode1",
                        "permissions": "read_write",
                      },
                    ],
                    "longview": Array [
                      Object {
                        "id": 5555,
                        "label": "longview1",
                        "permissions": "read_only",
                      },
                    ],
                    "nodebalancer": Array [
                      Object {
                        "id": 2222,
                        "label": "nb1",
                        "permissions": "read_only",
                      },
                    ],
                    "stackscript": Array [
                      Object {
                        "id": 4444,
                        "label": "stackscript1",
                        "permissions": null,
                      },
                    ],
                    "volume": Array [
                      Object {
                        "id": 8888,
                        "label": "volume1",
                        "permissions": null,
                      },
                    ],
                  },
                  "_polling": false,
                  "email": "example2@domain.com",
                  "password": "crisprcas9",
                  "restricted": true,
                  "username": "testuser2",
                },
              ],
            }
          }
        />,
        "history": Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "dispatch": [Function],
          "profile": Object {
            "authorized_keys": Array [
              "foobar",
            ],
            "email": "user@example.com",
            "email_notifications": true,
            "ip_whitelist_enabled": false,
            "lish_auth_method": "password_keys",
            "referrals": Object {
              "code": "",
              "completed": 0,
              "credit": 0,
              "pending": 0,
              "total": 0,
              "url": null,
            },
            "restricted": false,
            "timezone": "US/Eastern",
            "two_factor_auth": true,
            "username": "theuser",
          },
          "selectedMap": Object {},
          "users": Object {
            "ids": Array [
              undefined,
              undefined,
            ],
            "pagesFetched": Array [
              1,
            ],
            "plural": "users",
            "singular": "user",
            "totalPages": 1,
            "totalResults": 2,
            "users": Array [
              Object {
                "_polling": false,
                "email": "example1@domain.com",
                "password": "crisprcas9",
                "restricted": false,
                "username": "testuser1",
              },
              Object {
                "_permissions": Object {
                  "domain": Array [
                    Object {
                      "id": 3333,
                      "label": "domain1",
                      "permissions": "read_write",
                    },
                  ],
                  "global": Object {
                    "account_access": "read_only",
                    "add_domains": true,
                    "add_images": true,
                    "add_linodes": true,
                    "add_longview": true,
                    "add_nodebalancers": true,
                    "add_stackscripts": true,
                    "add_volumes": true,
                    "cancel_account": false,
                    "longview_subscription": true,
                  },
                  "image": Array [
                    Object {
                      "id": 6666,
                      "label": "image1",
                      "permissions": "read_only",
                    },
                    Object {
                      "id": 7777,
                      "label": "image2",
                      "permissions": "read_write",
                    },
                  ],
                  "linode": Array [
                    Object {
                      "id": 1111,
                      "label": "linode1",
                      "permissions": "read_write",
                    },
                  ],
                  "longview": Array [
                    Object {
                      "id": 5555,
                      "label": "longview1",
                      "permissions": "read_only",
                    },
                  ],
                  "nodebalancer": Array [
                    Object {
                      "id": 2222,
                      "label": "nb1",
                      "permissions": "read_only",
                    },
                  ],
                  "stackscript": Array [
                    Object {
                      "id": 4444,
                      "label": "stackscript1",
                      "permissions": null,
                    },
                  ],
                  "volume": Array [
                    Object {
                      "id": 8888,
                      "label": "volume1",
                      "permissions": null,
                    },
                  ],
                },
                "_polling": false,
                "email": "example2@domain.com",
                "password": "crisprcas9",
                "restricted": true,
                "username": "testuser2",
              },
            ],
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
